src/passes/Unsubtyping.cpp

```diff
@@ -0,0 +1,482 @@
+    // the relationships we need to keep all our type definitions valid.
+    while (!worklist.empty()) {
+      while (!worklist.empty()) {
+        auto it = worklist.begin();
+        auto type = *it;
+        worklist.erase(it);
+        auto super = supertypes.at(type);
+        if (super.isBasic()) {
+          continue;
+        }
+
+        if (type.isStruct()) {
+          const auto& fields = type.getStruct().fields;
+          const auto& superFields = super.getStruct().fields;
+          for (size_t i = 0, size = superFields.size(); i < size; ++i) {
+            if (fields[i].mutable_ == Immutable) {
+              noteSubtype(fields[i].type, superFields[i].type);
+            }
+          }
+        } else if (type.isArray()) {
+          auto elem = type.getArray().element;
+          if (elem.mutable_ == Immutable) {
+            noteSubtype(elem.type, super.getArray().element.type);
+          }
+        } else {
+          assert(type.isSignature());
+          auto sig = type.getSignature();
+          auto superSig = super.getSignature();
+          noteSubtype(superSig.params, sig.params);
+          noteSubtype(sig.results, superSig.results);
+        }
+      }
+
+      // TODO: This is expensive. Analyze casts incrementally after we
+      // initially analyze them.
+      analyzeCasts();
+    }
+  }
+
+  void analyzeCasts() {
+    // For each cast (src, dest) pair, any type that remains a subtype of src
+    // (meaning its values can inhabit locations typed src) and that was
+    // originally a subtype of dest (meaning its values would have passed the
+    // cast) should remain a subtype of dest so that its values continue to pass
+    // the cast.
+    //
+    // For every type, walk up its new supertype chain to find cast sources and
+    // compare against their associated cast destinations.
+    for (auto& [type, _] : typeDepths) {
+      for (auto superIt = supertypes.find(type); superIt != supertypes.end();
+           superIt = supertypes.find(superIt->second)) {
+        auto super = superIt->second;
+        auto destsIt = castTypes.find(super);
+        if (destsIt == castTypes.end()) {
+          continue;
+        }
+        for (auto dest : destsIt->second) {
+          if (HeapType::isSubType(type, dest)) {
+            noteSubtype(type, dest);
+          }
+        }
+      }
+    }
+  }
+
+  void optimizeTypes(Module& wasm) {
+    struct Rewriter : GlobalTypeRewriter {
+      Unsubtyping& parent;
+      Rewriter(Unsubtyping& parent, Module& wasm)
+        : GlobalTypeRewriter(wasm), parent(parent) {}
+      std::optional<HeapType> getSuperType(HeapType type) override {
+        if (auto it = parent.supertypes.find(type);
+            it != parent.supertypes.end() && !it->second.isBasic()) {
+          return it->second;
+        }
+        return std::nullopt;
+      }
+    };
+    Rewriter(*this, wasm).update();
+  }
+
+  void visitFunction(Function* func) {
```

Comment:
Perhaps all the visitors could be in a separate class? That would be easier to read and also make the code reusable in the future, I think. That is, this separate class would just issue lots of calls to `parent.noteSubtype` (or it could store the results in itself, and then we could run those in parallel and then merge them, to use multiple cores)

Reply:
This makes sense as a way to parallelize the pass, but I'm surprised that you think it would be easier to read if these were in a separate class. Why is that?

Reply:
Just as a separation of concerns. That other class would gather the raw data of what type is assigned to what, and nothing more. The user of the class would get that data and not care how it was gathered (all the visitors).

Reply:
As we discussed, the raw data is not that useful on its own because it is incomplete. Since we have to walk both function code and other module code, I found it easier to parallelize this pass without separating out the visitors.

Reply:
As an example of a use I'd like to make of this information, the validator could look at the accumulated results of `noteSubtype` and issue an error for content that doesn't fit in those constraints. But I guess we can leave that to later refactoring.

Reply:
Yeah, I think leaving it to a later refactoring would be good.